// @nxcode/sdk/dist/nxcode.esm.js
/**
 * Nxcode SDK - Authentication Module
 */
const STORAGE_KEY = 'nxcode_sdk_auth';
class NxcodeAuth {
    constructor(apiEndpoint, appId) {
        this.user = null;
        this.token = null;
        this.refreshToken = null;
        this.expiresAt = null;
        this.authCallbacks = [];
        this.popupWindow = null;
        this.messageHandler = null;
        this.apiEndpoint = apiEndpoint;
        this.appId = appId;
        this.loadFromStorage();
        this.setupMessageListener();
    }
    /**
     * Initiate login flow via OAuth popup
     */
    async login(provider = 'google') {
        return new Promise((resolve, reject) => {
            // Open popup
            const width = 500;
            const height = 600;
            const left = window.screenX + (window.outerWidth - width) / 2;
            const top = window.screenY + (window.outerHeight - height) / 2;
            const loginUrl = `${this.apiEndpoint}/api/sdk/auth/login?app_id=${this.appId}&provider=${provider}`;
            this.popupWindow = window.open(loginUrl, 'nxcode_login', `width=${width},height=${height},left=${left},top=${top}`);
            if (!this.popupWindow) {
                reject(new Error('Failed to open login popup. Please allow popups for this site.'));
                return;
            }
            // Set up timeout
            const timeout = setTimeout(() => {
                this.popupWindow?.close();
                reject(new Error('Login timeout'));
            }, 120000); // 2 minute timeout
            // Wait for message from popup
            const handleMessage = (event) => {
                if (event.data?.type !== 'NXCODE_SDK_AUTH')
                    return;
                clearTimeout(timeout);
                window.removeEventListener('message', handleMessage);
                const result = event.data.payload;
                if (result.success && result.user && result.token) {
                    this.setAuth(result);
                    resolve(result.user);
                }
                else {
                    reject(new Error(result.error || 'Login failed'));
                }
            };
            window.addEventListener('message', handleMessage);
        });
    }
    /**
     * Logout and clear session
     */
    async logout() {
        try {
            if (this.token) {
                await fetch(`${this.apiEndpoint}/api/sdk/auth/logout`, {
                    method: 'POST',
                    headers: this.getHeaders(),
                });
            }
        }
        catch (error) {
            console.warn('Logout request failed:', error);
        }
        this.clearAuth();
    }
    /**
     * Get current user
     */
    getUser() {
        return this.user;
    }
    /**
     * Get current auth token
     */
    getToken() {
        // Check if token is expired
        if (this.expiresAt && new Date() > this.expiresAt) {
            // Token is expired, need to refresh
            // Start refresh in background but don't return the expired token
            this.refreshSession().catch(console.error);
            // Return null to force re-auth rather than using expired token
            return null;
        }
        return this.token;
    }
    /**
     * Get token, refreshing if needed (async version)
     */
    async getValidToken() {
        if (this.expiresAt && new Date() > this.expiresAt) {
            await this.refreshSession();
        }
        return this.token;
    }
    /**
     * Register callback for auth state changes
     */
    onAuthStateChange(callback) {
        this.authCallbacks.push(callback);
        // Call immediately with current state
        callback(this.user);
        // Return unsubscribe function
        return () => {
            const index = this.authCallbacks.indexOf(callback);
            if (index > -1) {
                this.authCallbacks.splice(index, 1);
            }
        };
    }
    /**
     * Check if user is logged in
     */
    isLoggedIn() {
        return this.user !== null && this.token !== null;
    }
    /**
     * Refresh the current session
     */
    async refreshSession() {
        if (!this.refreshToken) {
            this.clearAuth();
            return;
        }
        try {
            const response = await fetch(`${this.apiEndpoint}/api/sdk/auth/refresh`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                    'X-App-Id': this.appId,
                },
                body: JSON.stringify({ refreshToken: this.refreshToken }),
            });
            if (!response.ok) {
                throw new Error('Refresh failed');
            }
            const data = await response.json();
            this.setAuth({
                success: true,
                token: data.token,
                refreshToken: data.refreshToken,
                expiresAt: data.expiresAt,
                user: data.user,
            });
        }
        catch (error) {
            console.error('Session refresh failed:', error);
            this.clearAuth();
        }
    }
    /**
     * Fetch current user info from server
     */
    async fetchUser() {
        if (!this.token)
            return null;
        try {
            const response = await fetch(`${this.apiEndpoint}/api/sdk/auth/me`, {
                headers: this.getHeaders(),
            });
            if (!response.ok) {
                if (response.status === 401) {
                    await this.refreshSession();
                    return this.user;
                }
                throw new Error('Failed to fetch user');
            }
            const user = await response.json();
            this.user = user;
            this.notifyAuthChange();
            this.saveToStorage();
            return user;
        }
        catch (error) {
            console.error('Failed to fetch user:', error);
            return null;
        }
    }
    // ==================== Private Methods ====================
    setAuth(result) {
        if (result.success && result.user && result.token) {
            this.user = result.user;
            this.token = result.token;
            this.refreshToken = result.refreshToken || null;
            this.expiresAt = result.expiresAt ? new Date(result.expiresAt) : null;
            this.saveToStorage();
            this.notifyAuthChange();
        }
    }
    clearAuth() {
        this.user = null;
        this.token = null;
        this.refreshToken = null;
        this.expiresAt = null;
        localStorage.removeItem(`${STORAGE_KEY}_${this.appId}`);
        this.notifyAuthChange();
    }
    saveToStorage() {
        if (this.user && this.token) {
            const data = {
                token: this.token,
                refreshToken: this.refreshToken || '',
                expiresAt: this.expiresAt?.toISOString() || '',
                user: this.user,
            };
            localStorage.setItem(`${STORAGE_KEY}_${this.appId}`, JSON.stringify(data));
        }
    }
    loadFromStorage() {
        try {
            const stored = localStorage.getItem(`${STORAGE_KEY}_${this.appId}`);
            if (stored) {
                const data = JSON.parse(stored);
                this.token = data.token;
                this.refreshToken = data.refreshToken;
                this.expiresAt = data.expiresAt ? new Date(data.expiresAt) : null;
                this.user = data.user;
                // Check if expired
                if (this.expiresAt && new Date() > this.expiresAt) {
                    this.refreshSession().catch(() => this.clearAuth());
                }
            }
        }
        catch (error) {
            console.warn('Failed to load auth from storage:', error);
        }
    }
    notifyAuthChange() {
        for (const callback of this.authCallbacks) {
            try {
                callback(this.user);
            }
            catch (error) {
                console.error('Auth callback error:', error);
            }
        }
    }
    setupMessageListener() {
        // Already set up message handling in login()
    }
    getHeaders() {
        const headers = {
            'Content-Type': 'application/json',
            'X-App-Id': this.appId,
        };
        if (this.token) {
            headers['Authorization'] = `Bearer ${this.token}`;
        }
        return headers;
    }
}

/**
 * Nxcode SDK - Billing Module
 */
class NxcodeBilling {
    constructor(apiEndpoint, appId, getToken) {
        this.apiEndpoint = apiEndpoint;
        this.appId = appId;
        this.getToken = getToken;
    }
    /**
     * Get user's current balance
     */
    async getBalance() {
        const token = this.getToken();
        if (!token) {
            throw new Error('Not authenticated. Please login first.');
        }
        const response = await fetch(`${this.apiEndpoint}/api/sdk/billing/balance`, {
            headers: this.getHeaders(token),
        });
        if (!response.ok) {
            if (response.status === 401) {
                throw new Error('Session expired. Please login again.');
            }
            throw new Error('Failed to fetch balance');
        }
        const data = await response.json();
        return data.balance;
    }
    /**
     * Open top-up page in new tab
     */
    topUp() {
        const topUpUrl = `${this.apiEndpoint}/api/sdk/billing/topup?app_id=${this.appId}`;
        window.open(topUpUrl, '_blank');
    }
    // ==================== Private Methods ====================
    getHeaders(token) {
        return {
            'Content-Type': 'application/json',
            'X-App-Id': this.appId,
            'Authorization': `Bearer ${token}`,
        };
    }
}

/**
 * Nxcode SDK - Payment Module
 */
class NxcodePayment {
    constructor(apiEndpoint, appId, getToken) {
        this.apiEndpoint = apiEndpoint;
        this.appId = appId;
        this.getToken = getToken;
    }
    /**
     * Process an in-app purchase
     *
     * @param options - Charge options
     * @returns Charge result with transaction details
     */
    async charge(options) {
        const token = this.getToken();
        if (!token) {
            return {
                success: false,
                error: 'Not authenticated. Please login first.',
            };
        }
        try {
            const response = await fetch(`${this.apiEndpoint}/api/sdk/payment/charge`, {
                method: 'POST',
                headers: this.getHeaders(token),
                body: JSON.stringify({
                    amount: options.amount,
                    description: options.description,
                    metadata: options.metadata,
                }),
            });
            if (!response.ok) {
                if (response.status === 401) {
                    return {
                        success: false,
                        error: 'Session expired. Please login again.',
                    };
                }
                if (response.status === 402) {
                    return {
                        success: false,
                        error: 'Insufficient balance. Please top up.',
                    };
                }
                const errorData = await response.json().catch(() => ({}));
                return {
                    success: false,
                    error: errorData.detail || 'Payment failed',
                };
            }
            const result = await response.json();
            return result;
        }
        catch (error) {
            return {
                success: false,
                error: error instanceof Error ? error.message : 'Payment failed',
            };
        }
    }
    /**
     * Get user's transaction history
     *
     * @param limit - Maximum number of transactions to return
     * @param offset - Pagination offset
     * @returns Array of transactions
     */
    async getTransactions(limit = 50, offset = 0) {
        const token = this.getToken();
        if (!token) {
            throw new Error('Not authenticated. Please login first.');
        }
        const response = await fetch(`${this.apiEndpoint}/api/sdk/payment/transactions?limit=${limit}&offset=${offset}`, {
            headers: this.getHeaders(token),
        });
        if (!response.ok) {
            if (response.status === 401) {
                throw new Error('Session expired. Please login again.');
            }
            throw new Error('Failed to fetch transactions');
        }
        const data = await response.json();
        return data.transactions;
    }
    // ==================== Private Methods ====================
    getHeaders(token) {
        return {
            'Content-Type': 'application/json',
            'X-App-Id': this.appId,
            'Authorization': `Bearer ${token}`,
        };
    }
}

/**
 * Nxcode SDK - AI Module
 *
 * Provides access to AI capabilities through the Nxcode AI Gateway.
 * All AI calls require user authentication and are billed to the user's balance.
 */
/**
 * Convert simplified schema to JSON Schema (Gemini format)
 */
function toJsonSchema(schema) {
    // Primitive types
    if (schema === 'string') {
        return { type: 'STRING' };
    }
    if (schema === 'number') {
        return { type: 'NUMBER' };
    }
    if (schema === 'boolean') {
        return { type: 'BOOLEAN' };
    }
    // Array type
    if (Array.isArray(schema)) {
        if (schema.length !== 1) {
            throw new Error('Array schema must have exactly one element type');
        }
        return {
            type: 'ARRAY',
            items: toJsonSchema(schema[0]),
        };
    }
    // Object type
    if (typeof schema === 'object' && schema !== null) {
        const properties = {};
        const required = [];
        for (const [key, value] of Object.entries(schema)) {
            properties[key] = toJsonSchema(value);
            required.push(key); // All fields required by default
        }
        return {
            type: 'OBJECT',
            properties,
            required,
        };
    }
    throw new Error(`Invalid schema type: ${schema}`);
}
class NxcodeAI {
    constructor(apiEndpoint, appId, getToken) {
        this.apiEndpoint = apiEndpoint;
        this.appId = appId;
        this.getToken = getToken;
    }
    /**
     * Send a chat message and get a response.
     *
     * @example
     * const response = await Nxcode.ai.chat({
     *   messages: [
     *     { role: 'user', content: 'Hello!' }
     *   ]
     * });
     * console.log(response.content);
     */
    async chat(options) {
        const token = this.getToken();
        // Allow anonymous requests for dev preview and apps with anonymous AI enabled
        const headers = token ? this.getHeaders(token) : this.getAnonymousHeaders();
        const model = options.model || 'fast';
        // Build Gemini API request
        const geminiRequest = {
            contents: this.convertMessagesToGemini(options.messages),
        };
        // Add structured output config if schema provided
        if (options.responseSchema) {
            geminiRequest.generationConfig = {
                responseMimeType: 'application/json',
                responseSchema: toJsonSchema(options.responseSchema),
            };
        }
        const response = await fetch(`${this.apiEndpoint}/api/ai-gateway/v1beta/models/${model}:generateContent`, {
            method: 'POST',
            headers,
            body: JSON.stringify(geminiRequest),
        });
        if (!response.ok) {
            await this.handleError(response);
        }
        const data = await response.json();
        return this.parseGeminiResponse(data);
    }
    /**
     * Generate text from a prompt.
     *
     * @example
     * const response = await Nxcode.ai.generate({
     *   prompt: 'Write a haiku about coding'
     * });
     * console.log(response.text);
     */
    async generate(options) {
        const result = await this.chat({
            messages: [{ role: 'user', content: options.prompt }],
            model: options.model,
            responseSchema: options.responseSchema,
        });
        return {
            text: result.content,
            usage: result.usage,
        };
    }
    /**
     * Stream a chat response in real-time.
     *
     * @example
     * await Nxcode.ai.chatStream({
     *   messages: [{ role: 'user', content: 'Tell me a story' }],
     *   onChunk: (chunk) => {
     *     process.stdout.write(chunk.content);
     *     if (chunk.done) console.log('\n--- Done ---');
     *   }
     * });
     */
    async chatStream(options) {
        const token = this.getToken();
        // Allow anonymous streaming if app permits
        const headers = token ? this.getHeaders(token) : this.getAnonymousHeaders();
        const model = options.model || 'fast';
        const geminiRequest = {
            contents: this.convertMessagesToGemini(options.messages),
        };
        const response = await fetch(`${this.apiEndpoint}/api/ai-gateway/v1beta/models/${model}:streamGenerateContent?alt=sse`, {
            method: 'POST',
            headers,
            body: JSON.stringify(geminiRequest),
        });
        if (!response.ok) {
            await this.handleError(response);
        }
        const reader = response.body?.getReader();
        if (!reader) {
            throw new Error('Streaming not supported');
        }
        const decoder = new TextDecoder();
        let buffer = '';
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                buffer += decoder.decode(value, { stream: true });
                const lines = buffer.split('\n');
                buffer = lines.pop() || '';
                for (const line of lines) {
                    if (line.startsWith('data: ')) {
                        const jsonStr = line.slice(6).trim();
                        if (jsonStr === '[DONE]') {
                            options.onChunk({ content: '', done: true });
                            return;
                        }
                        try {
                            const data = JSON.parse(jsonStr);
                            const chunk = this.parseStreamChunk(data);
                            options.onChunk(chunk);
                        }
                        catch {
                            // Skip invalid JSON
                        }
                    }
                }
            }
            // Process remaining buffer
            if (buffer.startsWith('data: ')) {
                const jsonStr = buffer.slice(6).trim();
                if (jsonStr && jsonStr !== '[DONE]') {
                    try {
                        const data = JSON.parse(jsonStr);
                        const chunk = this.parseStreamChunk(data);
                        options.onChunk(chunk);
                    }
                    catch {
                        // Skip invalid JSON
                    }
                }
            }
            options.onChunk({ content: '', done: true });
        }
        finally {
            reader.releaseLock();
        }
    }
    /**
     * Stream text generation from a prompt.
     *
     * @example
     * await Nxcode.ai.generateStream({
     *   prompt: 'Write a poem',
     *   onChunk: (chunk) => console.log(chunk.content)
     * });
     */
    async generateStream(options) {
        return this.chatStream({
            messages: [{ role: 'user', content: options.prompt }],
            model: options.model,
            onChunk: options.onChunk,
        });
    }
    // ==================== Private Methods ====================
    getHeaders(token) {
        return {
            'Content-Type': 'application/json',
            'X-App-Id': this.appId,
            Authorization: `Bearer ${token}`,
        };
    }
    getAnonymousHeaders() {
        return {
            'Content-Type': 'application/json',
            'X-App-Id': this.appId,
        };
    }
    convertMessagesToGemini(messages) {
        return messages.map((msg) => ({
            role: msg.role === 'assistant' ? 'model' : 'user',
            parts: this.convertContentToParts(msg.content),
        }));
    }
    convertContentToParts(content) {
        // Simple text content
        if (typeof content === 'string') {
            return [{ text: content }];
        }
        // Multimodal content
        return content.map((part) => {
            switch (part.type) {
                case 'text':
                    return { text: part.text };
                case 'image':
                    // Gemini format: inlineData with base64
                    return {
                        inlineData: {
                            mimeType: part.mimeType,
                            data: part.data,
                        },
                    };
                case 'image_url':
                    // Gemini format: fileData with URL
                    return {
                        fileData: {
                            fileUri: part.url,
                        },
                    };
                default:
                    return { text: '' };
            }
        });
    }
    parseGeminiResponse(data) {
        const candidates = data.candidates;
        const content = candidates?.[0]?.content?.parts?.[0]?.text || '';
        const usageMetadata = data.usageMetadata;
        return {
            content,
            usage: usageMetadata
                ? {
                    inputTokens: usageMetadata.promptTokenCount || 0,
                    outputTokens: usageMetadata.candidatesTokenCount || 0,
                }
                : undefined,
        };
    }
    parseStreamChunk(data) {
        const candidates = data.candidates;
        const content = candidates?.[0]?.content?.parts?.[0]?.text || '';
        const finishReason = candidates?.[0]?.finishReason;
        const done = finishReason === 'STOP' || finishReason === 'END_TURN';
        const usageMetadata = data.usageMetadata;
        return {
            content,
            done,
            usage: usageMetadata
                ? {
                    inputTokens: usageMetadata.promptTokenCount || 0,
                    outputTokens: usageMetadata.candidatesTokenCount || 0,
                }
                : undefined,
        };
    }
    async handleError(response) {
        let message = 'AI request failed';
        try {
            const data = await response.json();
            message = data.detail || data.message || message;
        }
        catch {
            // Ignore parse errors, use default message based on status
            if (response.status === 401) {
                message = 'Session expired. Please login again.';
            }
            else if (response.status === 402) {
                message = 'Insufficient balance. Please top up to continue.';
            }
        }
        throw new Error(message);
    }
}

/**
 * Nxcode SDK - Agent Module
 *
 * Provides AI agents with tool calling capabilities.
 * Agents can call backend endpoints as tools and maintain conversation history.
 */
// ==================== Helper Functions ====================
/**
 * Convert simple schema to JSON Schema format for Gemini
 */
function toJsonSchemaProperty(schema) {
    if (schema === 'string')
        return { type: 'STRING' };
    if (schema === 'number')
        return { type: 'NUMBER' };
    if (schema === 'boolean')
        return { type: 'BOOLEAN' };
    if (Array.isArray(schema)) {
        return {
            type: 'ARRAY',
            items: toJsonSchemaProperty(schema[0]),
        };
    }
    if (typeof schema === 'object') {
        const properties = {};
        const required = [];
        for (const [key, value] of Object.entries(schema)) {
            properties[key] = toJsonSchemaProperty(value);
            required.push(key);
        }
        return { type: 'OBJECT', properties, required };
    }
    return { type: 'STRING' };
}
/**
 * Convert AgentTool to Gemini function declaration
 */
function toolToGeminiFunction(tool) {
    const declaration = {
        name: tool.name,
        description: tool.description,
    };
    // Always include parameters object (even if empty) for Gemini compatibility
    const properties = {};
    const required = [];
    if (tool.parameters) {
        for (const [key, value] of Object.entries(tool.parameters)) {
            properties[key] = toJsonSchemaProperty(value);
            required.push(key);
        }
    }
    declaration.parameters = {
        type: 'OBJECT',
        properties,
        required,
    };
    return declaration;
}
// ==================== Agent Class ====================
class Agent {
    constructor(options, apiEndpoint, appId, getToken) {
        this.history = [];
        this.options = {
            instructions: options.instructions,
            tools: options.tools,
            model: options.model || 'fast',
            maxSteps: options.maxSteps || 10,
        };
        this.apiEndpoint = apiEndpoint;
        this.appId = appId;
        this.getToken = getToken;
    }
    /**
     * Run the agent with user input
     */
    async run(input, callbacks) {
        const token = this.getToken();
        // Add user message to history
        this.history.push({ role: 'user', content: input });
        const toolCalls = [];
        let totalInputTokens = 0;
        let totalOutputTokens = 0;
        let steps = 0;
        // Agent loop
        while (steps < this.options.maxSteps) {
            steps++;
            // Call Gemini with tools
            const response = await this.callGemini(token || undefined);
            totalInputTokens += response.usage?.inputTokens || 0;
            totalOutputTokens += response.usage?.outputTokens || 0;
            // Check if AI wants to call a function
            const functionCall = response.functionCall;
            if (functionCall) {
                callbacks?.onToolCall?.(functionCall.name, functionCall.args);
                // Find the tool
                const tool = this.options.tools.find((t) => t.name === functionCall.name);
                if (!tool) {
                    throw new Error(`Unknown tool: ${functionCall.name}`);
                }
                // Call the tool endpoint
                const result = await this.callTool(tool, functionCall.args, token || undefined);
                callbacks?.onToolResult?.(functionCall.name, result);
                // Record tool call
                toolCalls.push({
                    name: functionCall.name,
                    params: functionCall.args,
                    result,
                });
                // Add function call and response to history using raw parts from Gemini
                this.history.push({
                    role: 'model',
                    content: '',
                    parts: response.rawParts || [{ functionCall: { name: functionCall.name, args: functionCall.args } }],
                });
                this.history.push({
                    role: 'user',
                    content: '',
                    parts: [{ functionResponse: { name: functionCall.name, response: result } }],
                });
                // Continue loop to let AI process the result
                continue;
            }
            // No function call - AI is done, return the text response
            const output = response.text || '';
            // Add assistant response to history
            this.history.push({ role: 'assistant', content: output });
            return {
                output,
                toolCalls,
                usage: {
                    inputTokens: totalInputTokens,
                    outputTokens: totalOutputTokens,
                },
            };
        }
        throw new Error(`Agent exceeded maximum steps (${this.options.maxSteps})`);
    }
    /**
     * Clear conversation history
     */
    reset() {
        this.history = [];
    }
    /**
     * Get current conversation history
     */
    getHistory() {
        return [...this.history];
    }
    // ==================== Private Methods ====================
    async callGemini(token) {
        // Build contents from history
        const contents = this.history.map((msg) => {
            if (msg.parts) {
                return {
                    role: msg.role === 'assistant' ? 'model' : msg.role,
                    parts: msg.parts,
                };
            }
            return {
                role: msg.role === 'assistant' ? 'model' : 'user',
                parts: [{ text: msg.content }],
            };
        });
        // Build tools
        const tools = [
            {
                functionDeclarations: this.options.tools.map(toolToGeminiFunction),
            },
        ];
        // Build request with tool config to encourage tool use
        const request = {
            contents,
            tools,
            toolConfig: {
                functionCallingConfig: {
                    mode: 'AUTO', // Let model decide when to use tools
                },
            },
            systemInstruction: {
                parts: [{ text: this.options.instructions }],
            },
        };
        const response = await fetch(`${this.apiEndpoint}/api/ai-gateway/v1beta/models/${this.options.model}:generateContent`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'X-App-Id': this.appId,
                ...(token ? { Authorization: `Bearer ${token}` } : {}),
            },
            body: JSON.stringify(request),
        });
        if (!response.ok) {
            const error = await response.json().catch(() => ({}));
            throw new Error(error.detail || error.message || 'Agent API call failed');
        }
        const data = await response.json();
        // Parse response
        const candidate = data.candidates?.[0];
        const parts = candidate?.content?.parts || [];
        // Check for function call (support both camelCase and snake_case)
        for (const part of parts) {
            const funcCall = part.functionCall || part.function_call;
            if (funcCall) {
                return {
                    functionCall: {
                        name: funcCall.name,
                        args: funcCall.args || {},
                    },
                    rawParts: parts, // Save raw parts including thought_signature
                    usage: {
                        inputTokens: data.usageMetadata?.promptTokenCount || 0,
                        outputTokens: data.usageMetadata?.candidatesTokenCount || 0,
                    },
                };
            }
        }
        // Return text response
        const text = parts.find((p) => p.text)?.text || '';
        return {
            text,
            usage: {
                inputTokens: data.usageMetadata?.promptTokenCount || 0,
                outputTokens: data.usageMetadata?.candidatesTokenCount || 0,
            },
        };
    }
    async callTool(tool, params, token) {
        // Build context to pass to tool
        const body = {
            ...params,
            _context: {
                agentId: this.options.instructions.slice(0, 20), // Simple ID
                stepIndex: this.history.length,
                previousTools: this.history
                    .filter((m) => m.parts?.some((p) => p.functionResponse))
                    .map((m) => {
                    const fr = m.parts?.find((p) => p.functionResponse)?.functionResponse;
                    return fr ? { name: fr.name, result: fr.response } : null;
                })
                    .filter(Boolean),
            },
        };
        const response = await fetch(tool.endpoint, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                ...(token ? { Authorization: `Bearer ${token}` } : {}),
            },
            body: JSON.stringify(body),
        });
        if (!response.ok) {
            const error = await response.text();
            return { error: `Tool call failed: ${error}` };
        }
        return response.json();
    }
}

/**
 * Nxcode SDK
 *
 * A lightweight SDK for integrating Nxcode authentication, billing,
 * and payment features into web applications.
 *
 * Usage:
 *   <script src="https://sdk.nxcode.io/v1/nxcode.js"></script>
 *   <script>
 *     const user = await Nxcode.auth.login();
 *     console.log('Logged in:', user.email);
 *   </script>
 */
// API endpoint - auto-detect or use default
const DEFAULT_API_ENDPOINT = 'https://studio-api.nxcode.io';
class NxcodeSDK {
    constructor() {
        this.config = null;
        this._auth = null;
        this._billing = null;
        this._payment = null;
        this._ai = null;
        this.initPromise = null;
        this.apiEndpoint = this.detectApiEndpoint();
        this.autoInit();
    }
    /**
     * Auto-initialize SDK based on environment
     */
    async autoInit() {
        if (this.initPromise)
            return this.initPromise;
        this.initPromise = this.init();
        return this.initPromise;
    }
    /**
     * Initialize SDK by fetching configuration
     */
    async init() {
        try {
            // Try to get config from server
            const response = await fetch(`${this.apiEndpoint}/api/sdk/config`, {
                credentials: 'include',
            });
            if (response.ok) {
                this.config = await response.json();
                this.config.apiEndpoint = this.apiEndpoint;
                this.setupModules();
            }
        }
        catch (error) {
            console.warn('Nxcode SDK: Could not auto-detect app config. Call Nxcode.configure() manually.');
        }
    }
    /**
     * Manually configure SDK with app ID and optional endpoint
     */
    configure(appId, options) {
        if (options?.apiEndpoint) {
            this.apiEndpoint = options.apiEndpoint;
        }
        this.config = {
            appId,
            name: '',
            billingMode: 'creator_pays',
            apiEndpoint: this.apiEndpoint,
        };
        this.setupModules();
    }
    /**
     * Set up SDK modules after configuration
     */
    setupModules() {
        if (!this.config)
            return;
        this._auth = new NxcodeAuth(this.config.apiEndpoint, this.config.appId);
        this._billing = new NxcodeBilling(this.config.apiEndpoint, this.config.appId, () => this._auth?.getToken() || null);
        this._payment = new NxcodePayment(this.config.apiEndpoint, this.config.appId, () => this._auth?.getToken() || null);
        this._ai = new NxcodeAI(this.config.apiEndpoint, this.config.appId, () => this._auth?.getToken() || null);
    }
    /**
     * Detect API endpoint based on environment
     */
    detectApiEndpoint() {
        if (typeof window !== 'undefined') {
            const hostname = window.location.hostname;
            // Development environment
            if (hostname === 'localhost' || hostname === '127.0.0.1') {
                return 'http://localhost:8001';
            }
            // Check for dev environment
            if (hostname.includes('.nxcode.dev') || hostname.includes('.nxcode.app')) {
                return 'https://studio-api.nxcode.io';
            }
        }
        return DEFAULT_API_ENDPOINT;
    }
    /**
     * Ensure SDK is initialized before operations
     */
    async ensureInitialized() {
        if (this.initPromise) {
            await this.initPromise;
        }
        if (!this.config) {
            throw new Error('Nxcode SDK not configured. Add X-App-Id header or call Nxcode.configure(appId).');
        }
    }
    // ==================== Auth Module ====================
    get auth() {
        const self = this;
        return {
            /**
             * Login via OAuth popup
             */
            async login(provider = 'google') {
                await self.ensureInitialized();
                return self._auth.login(provider);
            },
            /**
             * Logout current user
             */
            async logout() {
                await self.ensureInitialized();
                return self._auth.logout();
            },
            /**
             * Get current user
             */
            getUser() {
                return self._auth?.getUser() || null;
            },
            /**
             * Get auth token for API calls
             */
            getToken() {
                return self._auth?.getToken() || null;
            },
            /**
             * Register auth state change callback
             */
            onAuthStateChange(callback) {
                if (!self._auth) {
                    // Store the actual unsubscribe function when SDK initializes
                    let actualUnsubscribe = null;
                    let cancelled = false;
                    // Queue callback until initialized
                    self.ensureInitialized().then(() => {
                        if (!cancelled && self._auth) {
                            actualUnsubscribe = self._auth.onAuthStateChange(callback);
                        }
                    });
                    // Return unsubscribe that handles both immediate and deferred cases
                    return () => {
                        cancelled = true;
                        if (actualUnsubscribe) {
                            actualUnsubscribe();
                        }
                    };
                }
                return self._auth.onAuthStateChange(callback);
            },
            /**
             * Check if user is logged in
             */
            isLoggedIn() {
                return self._auth?.isLoggedIn() || false;
            },
        };
    }
    // ==================== Billing Module ====================
    get billing() {
        const self = this;
        return {
            /**
             * Get user's current balance
             */
            async getBalance() {
                await self.ensureInitialized();
                return self._billing.getBalance();
            },
            /**
             * Open top-up page
             */
            topUp() {
                self._billing?.topUp();
            },
        };
    }
    // ==================== Payment Module ====================
    get payment() {
        const self = this;
        return {
            /**
             * Process in-app purchase
             */
            async charge(options) {
                await self.ensureInitialized();
                return self._payment.charge(options);
            },
            /**
             * Get transaction history
             */
            async getTransactions(limit, offset) {
                await self.ensureInitialized();
                return self._payment.getTransactions(limit, offset);
            },
        };
    }
    // ==================== AI Module ====================
    get ai() {
        const self = this;
        return {
            /**
             * Send a chat message and get a response
             *
             * @example
             * const response = await Nxcode.ai.chat({
             *   messages: [{ role: 'user', content: 'Hello!' }]
             * });
             */
            async chat(options) {
                await self.ensureInitialized();
                return self._ai.chat(options);
            },
            /**
             * Generate text from a prompt
             *
             * @example
             * const response = await Nxcode.ai.generate({ prompt: 'Write a haiku' });
             */
            async generate(options) {
                await self.ensureInitialized();
                return self._ai.generate(options);
            },
            /**
             * Stream a chat response in real-time
             *
             * @example
             * await Nxcode.ai.chatStream({
             *   messages: [{ role: 'user', content: 'Tell me a story' }],
             *   onChunk: (chunk) => {
             *     document.body.innerText += chunk.content;
             *     if (chunk.done) console.log('Done!');
             *   }
             * });
             */
            async chatStream(options) {
                await self.ensureInitialized();
                return self._ai.chatStream(options);
            },
            /**
             * Stream text generation from a prompt
             *
             * @example
             * await Nxcode.ai.generateStream({
             *   prompt: 'Write a poem',
             *   onChunk: (chunk) => console.log(chunk.content)
             * });
             */
            async generateStream(options) {
                await self.ensureInitialized();
                return self._ai.generateStream(options);
            },
            /**
             * Create an AI agent with tool calling capabilities
             *
             * @example
             * const agent = Nxcode.ai.createAgent({
             *   instructions: 'You are a helpful assistant that can search and send emails.',
             *   tools: [
             *     {
             *       name: 'search_emails',
             *       description: 'Search user emails',
             *       parameters: { query: 'string' },
             *       endpoint: '/api/emails/search'
             *     },
             *     {
             *       name: 'send_email',
             *       description: 'Send an email',
             *       parameters: { to: 'string', subject: 'string', body: 'string' },
             *       endpoint: '/api/emails/send'
             *     }
             *   ]
             * });
             *
             * const result = await agent.run('Find emails from John and reply to him');
             * console.log(result.output);
             */
            createAgent(options) {
                if (!self.config) {
                    throw new Error('Nxcode SDK not configured. Call Nxcode.configure(appId) first.');
                }
                return new Agent(options, self.apiEndpoint, self.config.appId, () => self._auth?.getToken() || null);
            },
        };
    }
    // ==================== Utility Methods ====================
    /**
     * Get current configuration
     */
    getConfig() {
        return this.config;
    }
    /**
     * Check if SDK is ready
     */
    isReady() {
        return this.config !== null && this._auth !== null;
    }
    /**
     * Wait for SDK to be ready
     */
    async ready() {
        await this.ensureInitialized();
    }
}
// Create singleton instance
const Nxcode = new NxcodeSDK();
// Export for UMD (browser global)
if (typeof window !== 'undefined') {
    window.Nxcode = Nxcode;
}

export { Agent, Nxcode, Nxcode as default };
